File: python/soma_client/open_duck_policy.py
from __future__ import annotations

import argparse
import json
import math
import queue
import struct
import time
from pathlib import Path

import numpy as np
import onnxruntime as ort
import zenoh

STATE_KEY = "soma/open-duck-v2/state"
TARGET_KEY = "soma/open-duck-v2/target"
STATE = struct.Struct("<23QII39f")
TARGET = struct.Struct("<4Q14f")
REJECTION_NAMES = ("decode", "timeline", "sequence", "expired", "invalid", "runtime_generation")
GAIT_PHASE_PERIOD = 27
OBSERVATION_SHAPE = (1, 101)
ACTION_SHAPE = (1, 14)
ACTION_SCALE = np.float32(0.25)
SLEW_RATE_RAD_S = np.float32(5.24)
POLICY_PERIOD_S = np.float32(0.02)
TARGET_TTL_NS = 40_000_000
TARGET_ABS_LIMIT_RAD = np.float32(10.0)
CHECKPOINT_SHA256 = "cb61453a8bcb547ccfdeb4f03ba0fa67ebcf767dcf4aa6e5c9a0d92b302f9b23"
DEFAULT_POSE = np.array([
    0.002, 0.053, -0.63, 1.368, -0.784, 0.0, 0.0, 0.0, 0.0,
    -0.003, -0.065, 0.635, 1.379, -0.796,
], dtype=np.float32)


def decode_state(payload: bytes) -> dict[str, object]:
    values = STATE.unpack(payload)
    facts = values[25:]
    reason = ("none", *REJECTION_NAMES)[values[23]] if values[23] <= 6 else "unknown"
    return {
        "sequence": values[0], "timeline": values[1], "capture_ns": values[2],
        "requested": values[3], "admitted": values[4], "applied": values[5],
        "message_age_ns": values[6], "runtime_dropped_targets": values[7],
        "rejection_counts": dict(zip(REJECTION_NAMES, values[8:14])),
        "max_rejection_age_ns": dict(zip(REJECTION_NAMES, values[14:20])),
        "last_rejection": {
            "reason": reason, "sequence": values[20], "age_ns": values[21],
            "ttl_ns": values[22],
        },
        "flags": values[24],
        "positions": np.asarray(facts[:14], dtype=np.float32),
        "velocities": np.asarray(facts[14:28], dtype=np.float32),
        "gyro": np.asarray(facts[28:31], dtype=np.float32),
        "acceleration": np.asarray(facts[31:34], dtype=np.float32),
        "contacts": np.asarray(facts[34:36], dtype=np.float32),
        "root_height": facts[36], "root_roll": facts[37], "root_pitch": facts[38],
    }


class LatestStateBuffer:
    def __init__(self) -> None:
        self.states: queue.Queue[bytes] = queue.Queue(maxsize=1)
        self.dropped = 0

    def receive(self, sample: object) -> None:
        try:
            self.states.get_nowait()
            self.dropped += 1
        except queue.Empty:
            pass
        try:
            self.states.put_nowait(bytes(sample.payload))
        except queue.Full:
            self.dropped += 1

    def get(self, timeout: float) -> bytes:
        return self.states.get(timeout=timeout)


class Policy:
    def __init__(self, checkpoint: Path, velocity_x: float = 0.3) -> None:
        options = ort.SessionOptions()
        options.intra_op_num_threads = 1
        options.inter_op_num_threads = 1
        options.execution_mode = ort.ExecutionMode.ORT_SEQUENTIAL
        self.session = ort.InferenceSession(str(checkpoint), sess_options=options,
                                            providers=["CPUExecutionProvider"])
        if tuple(self.session.get_inputs()[0].shape) != OBSERVATION_SHAPE:
            raise ValueError(f"Open Duck observation ABI mismatch: {self.session.get_inputs()[0].shape}")
        if tuple(self.session.get_outputs()[0].shape) != ACTION_SHAPE:
            raise ValueError(f"Open Duck action ABI mismatch: {self.session.get_outputs()[0].shape}")
        self.velocity_x = velocity_x
        self.default = DEFAULT_POSE.copy()
        self.previous = DEFAULT_POSE.copy()
        self.history = [np.zeros(14, np.float32) for _ in range(3)]
        self.phase_tick = 0
        self.phase = np.zeros(2, np.float32)
        self.last_observation: np.ndarray | None = None
        self.last_action: np.ndarray | None = None

    def infer(self, state: dict[str, object]) -> np.ndarray:
        required = ("positions", "velocities", "gyro", "acceleration", "contacts")
        if any(key not in state for key in required):
            raise ValueError("Open Duck state is missing required facts")
        positions = np.asarray(state["positions"], np.float32)
        velocities = np.asarray(state["velocities"], np.float32)
        gyro = np.asarray(state["gyro"], np.float32)
        contacts = np.asarray(state["contacts"], np.float32)
        if (positions.shape, velocities.shape, gyro.shape, contacts.shape) != ((14,), (14,), (3,), (2,)):
            raise ValueError("Open Duck state fact shape mismatch")
        if not all(np.isfinite(values).all() for values in (positions, velocities, gyro, contacts)):
            raise ValueError("Open Duck state contains non-finite values")
        acceleration = np.asarray(state["acceleration"], np.float32).copy()
        if acceleration.shape != (3,) or not np.isfinite(acceleration).all():
            raise ValueError("Open Duck acceleration shape or finite-value mismatch")
        acceleration[0] += 1.3
        command = np.array([self.velocity_x, 0, 0, 0, 0, 0, 0], np.float32)
        observation = np.concatenate((gyro, acceleration, command,
            positions - self.default, velocities * 0.05,
            *self.history, self.previous, np.asarray(state["contacts"], np.float32),
            self.phase)).astype(np.float32)
        if observation.shape != (101,) or not np.isfinite(observation).all():
            raise RuntimeError("invalid Open Duck policy observation or action")
        raw = np.asarray(self.session.run(None, {"obs": observation[None, :]})[0])
        if raw.shape != ACTION_SHAPE or not np.isfinite(raw).all():
            raise RuntimeError("invalid Open Duck policy action")
        action = raw[0].astype(np.float32)
        self.history = [action, *self.history[:2]]
        self.last_observation = observation.copy()
        self.last_action = action.copy()
        proposed = self.default + action * ACTION_SCALE
        self.previous = np.clip(proposed, self.previous - SLEW_RATE_RAD_S * POLICY_PERIOD_S,
                                self.previous + SLEW_RATE_RAD_S * POLICY_PERIOD_S)
        self.phase_tick = (self.phase_tick + 1) % GAIT_PHASE_PERIOD
        self.phase = np.array([
            math.cos(self.phase_tick / GAIT_PHASE_PERIOD * 2 * math.pi),
            math.sin(self.phase_tick / GAIT_PHASE_PERIOD * 2 * math.pi),
        ], np.float32)
        return self.previous

    @staticmethod
    def target_payload(sequence: int, state: dict[str, object], positions: np.ndarray) -> bytes:
        positions = np.asarray(positions, np.float32)
        if positions.shape != (14,) or not np.isfinite(positions).all():
            raise ValueError("Open Duck target shape or finite-value mismatch")
        if np.any(np.abs(positions) > TARGET_ABS_LIMIT_RAD):
            raise ValueError("Open Duck target is out of range")
        if "timeline" not in state or "capture_ns" not in state:
            raise ValueError("Open Duck target is missing lineage")
        return TARGET.pack(sequence, int(state["timeline"]), int(state["capture_ns"]),
                           TARGET_TTL_NS, *positions)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--checkpoint", type=Path, required=True)
    parser.add_argument("--stall-after", type=int)
    parser.add_argument("--duration", type=float)
    parser.add_argument("--ready-file", type=Path)
    parser.add_argument("--vx", type=float, default=0.3)
    args = parser.parse_args()
    config = zenoh.Config.from_json5('{mode:"client",connect:{endpoints:["tcp/127.0.0.1:7448"]},scouting:{multicast:{enabled:false}}}')
    policy = Policy(args.checkpoint, velocity_x=args.vx)
    latest = LatestStateBuffer()
    with zenoh.open(config) as session, session.declare_subscriber(STATE_KEY, latest.receive):
        if args.ready_file is not None:
            args.ready_file.touch()
        emitted = 0
        started = time.monotonic()
        evidence = {"states": 0, "applied": False, "expiry": False, "rejected": False,
                    "max_message_age_ns": 0, "last_requested": 0,
                    "last_admitted": 0, "last_applied": 0,
                    "min_root_height_m": float("inf"), "max_abs_roll_rad": 0.0,
                    "max_abs_pitch_rad": 0.0, "first_state_sequence": 0,
                    "last_state_sequence": 0, "max_state_sequence_gap": 0,
                    "max_inference_ns": 0, "dropped_states": 0,
                    "total_inference_ns": 0, "mean_inference_ns": 0,
                    "runtime_dropped_targets": 0,
                    "rejection_counts": {name: 0 for name in REJECTION_NAMES},
                    "max_rejection_age_ns": {name: 0 for name in REJECTION_NAMES},
                    "last_rejection": {"reason": "none", "sequence": 0,
                                       "age_ns": 0, "ttl_ns": 0}}
        while True:
            state = decode_state(latest.get(timeout=5))
            evidence["states"] += 1
            evidence["dropped_states"] = latest.dropped
            evidence["runtime_dropped_targets"] = int(state["runtime_dropped_targets"])
            sequence = int(state["sequence"])
            if evidence["first_state_sequence"] == 0:
                evidence["first_state_sequence"] = sequence
            if evidence["last_state_sequence"]:
                evidence["max_state_sequence_gap"] = max(
                    evidence["max_state_sequence_gap"],
                    sequence - evidence["last_state_sequence"])
            evidence["last_state_sequence"] = sequence
            evidence["applied"] |= bool(int(state["flags"]) & 1)
            evidence["expiry"] |= bool(int(state["flags"]) & 4)
            evidence["rejected"] |= bool(int(state["flags"]) & 8) or any(
                state["rejection_counts"].values())
            evidence["rejection_counts"] = {
                name: max(evidence["rejection_counts"][name], state["rejection_counts"][name])
                for name in REJECTION_NAMES
            }
            evidence["max_rejection_age_ns"] = {
                name: max(evidence["max_rejection_age_ns"][name], state["max_rejection_age_ns"][name])
                for name in REJECTION_NAMES
            }
            if state["last_rejection"]["reason"] != "none":
                evidence["last_rejection"] = state["last_rejection"]
            evidence["max_message_age_ns"] = max(evidence["max_message_age_ns"], int(state["message_age_ns"]))
            evidence["last_requested"] = int(state["requested"])
            evidence["last_admitted"] = int(state["admitted"])
            evidence["last_applied"] = int(state["applied"])
            evidence["min_root_height_m"] = min(evidence["min_root_height_m"], float(state["root_height"]))
            evidence["max_abs_roll_rad"] = max(evidence["max_abs_roll_rad"], abs(float(state["root_roll"])))
            evidence["max_abs_pitch_rad"] = max(evidence["max_abs_pitch_rad"], abs(float(state["root_pitch"])))
            if args.stall_after is not None and emitted >= args.stall_after:
                if evidence["applied"]:
                    if (args.duration is not None and time.monotonic() - started >= args.duration
                            and evidence["expiry"]):
                        print(json.dumps({"status": "stall-complete", "emitted": emitted, **evidence}))
                        return
                    continue
            inference_started = time.monotonic_ns()
            target = policy.infer(state)
            inference_ns = time.monotonic_ns() - inference_started
            evidence["max_inference_ns"] = max(evidence["max_inference_ns"], inference_ns)
            evidence["total_inference_ns"] += inference_ns
            emitted += 1
            evidence["mean_inference_ns"] = evidence["total_inference_ns"] // emitted
            payload = policy.target_payload(int(state["sequence"]), state, target)
            session.put(TARGET_KEY, payload)
            if (args.duration is not None and time.monotonic() - started >= args.duration
                    and evidence["applied"]):
                print(json.dumps({"status": "complete", "emitted": emitted, **evidence}))
                return


if __name__ == "__main__":
    main()

File: crates/soma-runtime/src/bin/open-duck-policy.rs
use ndarray::Array2;
use ort::{
    session::Session,
    tensor::TensorElementType,
    value::{TensorRef, ValueType},
};
use sha2::{Digest, Sha256};
use soma_runtime::open_duck::{
    decode_state, encode_target, OpenDuckPolicy, OPEN_DUCK_OBSERVATION, OPEN_DUCK_STATE_KEY,
    OPEN_DUCK_TARGET_BYTES, OPEN_DUCK_TARGET_KEY,
};
use std::{
    env, fs,
    path::PathBuf,
    time::{Duration, Instant},
};
use zenoh::Config;

const CHECKPOINT_SHA256: &str = "cb61453a8bcb547ccfdeb4f03ba0fa67ebcf767dcf4aa6e5c9a0d92b302f9b23";

struct Args {
    checkpoint: PathBuf,
    duration: Option<f64>,
    ready_file: Option<PathBuf>,
    vx: f32,
    stall_after: Option<u64>,
    parity_fixture: Option<PathBuf>,
}

fn args() -> Result<Args, String> {
    let mut checkpoint = None;
    let mut duration = None;
    let mut ready_file = None;
    let mut vx = 0.3;
    let mut stall_after = None;
    let mut parity_fixture = None;
    let mut it = env::args().skip(1);
    while let Some(arg) = it.next() {
        let mut value = || it.next().ok_or_else(|| format!("missing value for {arg}"));
        match arg.as_str() {
            "--checkpoint" => checkpoint = Some(PathBuf::from(value()?)),
            "--duration" => duration = Some(value()?.parse().map_err(|_| "invalid duration")?),
            "--ready-file" => ready_file = Some(PathBuf::from(value()?)),
            "--vx" => vx = value()?.parse().map_err(|_| "invalid vx")?,
            "--stall-after" => stall_after = Some(value()?.parse().map_err(|_| "invalid stall-after")?),
            "--parity-fixture" => parity_fixture = Some(PathBuf::from(value()?)),
            "-h" | "--help" => return Err("usage: open-duck-policy --checkpoint PATH [--duration SEC] [--ready-file PATH] [--vx N] [--stall-after N]".into()),
            other => return Err(format!("unknown argument {other}")),
        }
    }
    Ok(Args {
        checkpoint: checkpoint.ok_or("--checkpoint is required")?,
        duration,
        ready_file,
        vx,
        stall_after,
        parity_fixture,
    })
}

fn verify_checkpoint(path: &PathBuf) -> Result<(), String> {
    let bytes = fs::read(path).map_err(|e| format!("read checkpoint: {e}"))?;
    let expected =
        env::var("SOMA_OPEN_DUCK_CHECKPOINT_SHA256").unwrap_or_else(|_| CHECKPOINT_SHA256.into());
    verify_checkpoint_digest(&bytes, &expected)
}

fn verify_checkpoint_digest(bytes: &[u8], expected: &str) -> Result<(), String> {
    let digest = format!("{:x}", Sha256::digest(bytes));
    if digest != expected {
        return Err(format!(
            "checkpoint sha256 mismatch: got {digest}, expected {expected}"
        ));
    }
    Ok(())
}

fn infer(
    session: &mut Session,
    observation: &[f32; OPEN_DUCK_OBSERVATION],
) -> Result<[f32; 14], String> {
    let input = Array2::from_shape_vec((1, OPEN_DUCK_OBSERVATION), observation.to_vec())
        .map_err(|e| e.to_string())?;
    let outputs = session
        .run(ort::inputs!["obs" => TensorRef::from_array_view(&input).map_err(|e| e.to_string())?])
        .map_err(|e| e.to_string())?;
    let (_, values) = outputs[0]
        .try_extract_tensor::<f32>()
        .map_err(|e| e.to_string())?;
    if values.len() != 14 || !values.iter().all(|v| v.is_finite()) {
        return Err("policy output ABI or finite-value mismatch".into());
    }
    values
        .try_into()
        .map_err(|_| "policy output shape mismatch".into())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let args = args().map_err(|e| format!("argument error: {e}"))?;
    verify_checkpoint(&args.checkpoint)?;
    let ort_path = env::var_os("ORT_DYLIB_PATH")
        .ok_or("ORT_DYLIB_PATH must point to the provisioned native ONNX Runtime")?;
    ort::init_from(&ort_path)
        .map_err(|e| {
            format!(
                "load ONNX Runtime from {}: {e}",
                PathBuf::from(&ort_path).display()
            )
        })?
        .commit();
    let mut session = Session::builder()?
        .with_intra_threads(1)?
        .with_inter_threads(1)?
        .commit_from_file(&args.checkpoint)?;
    let input_ok = session.inputs().first().is_some_and(|input| {
        input.name() == "obs"
            && matches!(input.dtype(), ValueType::Tensor { ty: TensorElementType::Float32, shape, .. } if shape.as_ref() == [1, 101])
    });
    let output_ok = session.outputs().first().is_some_and(|output| {
        matches!(output.dtype(), ValueType::Tensor { ty: TensorElementType::Float32, shape, .. } if shape.as_ref() == [1, 14])
    });
    if session.inputs().len() != 1 || session.outputs().len() != 1 || !input_ok || !output_ok {
        return Err("Open Duck model input/output ABI mismatch".into());
    }
    infer(&mut session, &[0.0; OPEN_DUCK_OBSERVATION])
        .map_err(|e| format!("Open Duck model warm-up failed: {e}"))?;
    if let Some(path) = &args.parity_fixture {
        let fixture: serde_json::Value = serde_json::from_slice(&fs::read(path)?)?;
        let values = fixture["observation"]
            .as_array()
            .ok_or("parity fixture is missing observation")?;
        let observation: [f32; OPEN_DUCK_OBSERVATION] = values
            .iter()
            .map(|value| value.as_f64().map(|value| value as f32))
            .collect::<Option<Vec<_>>>()
            .ok_or("parity fixture observation is not numeric")?
            .try_into()
            .map_err(|_| "parity fixture observation width mismatch")?;
        println!(
            "{}",
            serde_json::to_string(&infer(&mut session, &observation)?)?
        );
        return Ok(());
    }
    let zenoh = zenoh::open(Config::from_json5(r#"{mode:"client",connect:{endpoints:["tcp/127.0.0.1:7448"]},scouting:{multicast:{enabled:false}}}"#)?).await?;
    let subscriber = zenoh.declare_subscriber(OPEN_DUCK_STATE_KEY).await?;
    let publisher = zenoh.declare_publisher(OPEN_DUCK_TARGET_KEY).await?;
    if let Some(path) = &args.ready_file {
        fs::write(path, b"ready\n")?;
    }
    let started = Instant::now();
    let mut policy = OpenDuckPolicy::default();
    let mut emitted = 0_u64;
    let mut states = 0_u64;
    let mut dropped = 0_u64;
    let mut applied = false;
    let mut expiry = false;
    let mut rejected = false;
    let mut max_message_age_ns = 0_u64;
    let mut min_root_height_m = f32::INFINITY;
    let mut max_abs_roll_rad = 0_f32;
    let mut max_abs_pitch_rad = 0_f32;
    let mut first_state_sequence = 0_u64;
    let mut last_state_sequence = 0_u64;
    let mut max_state_sequence_gap = 0_u64;
    let mut max_inference_ns = 0_u64;
    let mut total_inference_ns = 0_u64;
    let mut payload = [0_u8; OPEN_DUCK_TARGET_BYTES];
    loop {
        let sample = subscriber.recv_async().await?;
        let mut latest = sample.payload().to_bytes().to_vec();
        while let Some(sample) = subscriber.try_recv()? {
            dropped = dropped.saturating_add(1);
            latest = sample.payload().to_bytes().to_vec();
        }
        let state = decode_state(&latest).ok_or("invalid Open Duck state payload")?;
        states += 1;
        if first_state_sequence == 0 {
            first_state_sequence = state.sequence;
        }
        if last_state_sequence != 0 {
            max_state_sequence_gap =
                max_state_sequence_gap.max(state.sequence.saturating_sub(last_state_sequence));
        }
        last_state_sequence = state.sequence;
        applied |= state.flags & 1 != 0;
        expiry |= state.flags & 4 != 0;
        rejected |= state.flags & 8 != 0;
        max_message_age_ns = max_message_age_ns.max(state.message_age_ns);
        min_root_height_m = min_root_height_m.min(state.root_height_m);
        max_abs_roll_rad = max_abs_roll_rad.max(state.root_roll_rad.abs());
        max_abs_pitch_rad = max_abs_pitch_rad.max(state.root_pitch_rad.abs());
        if args.stall_after.is_none_or(|limit| emitted < limit) {
            let observation = policy
                .observation(&state, args.vx)
                .ok_or("invalid policy observation")?;
            let inference_started = Instant::now();
            let action = infer(&mut session, &observation)?;
            let inference_ns = inference_started.elapsed().as_nanos() as u64;
            max_inference_ns = max_inference_ns.max(inference_ns);
            total_inference_ns = total_inference_ns.saturating_add(inference_ns);
            let target = policy
                .apply_action(action, &state)
                .ok_or("invalid policy target")?;
            encode_target(&target, &mut payload);
            publisher.put(payload.to_vec()).await?;
            emitted += 1;
        }
        let elapsed = args
            .duration
            .is_some_and(|limit| started.elapsed() >= Duration::from_secs_f64(limit));
        let stall_done = args.stall_after.is_some() && elapsed && applied && expiry;
        if (args.stall_after.is_none() && elapsed && applied) || stall_done {
            let status = if stall_done {
                "stall-complete"
            } else {
                "complete"
            };
            println!(
                concat!(
                "{{\"status\":\"{}\",\"emitted\":{},\"states\":{},\"applied\":{},\"expiry\":{},",
                "\"rejected\":{},\"max_message_age_ns\":{},\"last_requested\":{},",
                "\"last_admitted\":{},\"last_applied\":{},\"min_root_height_m\":{},",
                "\"max_abs_roll_rad\":{},\"max_abs_pitch_rad\":{},\"first_state_sequence\":{},",
                "\"last_state_sequence\":{},\"max_state_sequence_gap\":{},\"max_inference_ns\":{},",
                "\"mean_inference_ns\":{},\"dropped_states\":{},\"runtime_dropped_targets\":{}}}"),
                status,
                emitted,
                states,
                applied,
                expiry,
                rejected,
                max_message_age_ns,
                state.requested_sequence,
                state.admitted_sequence,
                state.applied_sequence,
                min_root_height_m,
                max_abs_roll_rad,
                max_abs_pitch_rad,
                first_state_sequence,
                last_state_sequence,
                max_state_sequence_gap,
                max_inference_ns,
                total_inference_ns / emitted.max(1),
                dropped,
                state.runtime_dropped_targets
            );
            return Ok(());
        }
    }
}

#[cfg(test)]
mod tests {
    use super::verify_checkpoint_digest;

    #[test]
    fn checkpoint_digest_is_pinned_and_mismatch_is_attributable() {
        let digest = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad";
        assert!(verify_checkpoint_digest(b"abc", digest).is_ok());
        let error = verify_checkpoint_digest(b"changed", digest).unwrap_err();
        assert!(error.contains("checkpoint sha256 mismatch"));
        assert!(error.contains(digest));
    }
}
